deploy/ciao-deploy-server/main.go

```diff
@@ -0,0 +1,476 @@
+	"path"
+	"sync"
+	"syscall"
+
+	"encoding/pem"
+
+	"crypto/x509"
+
+	"path/filepath"
+
+	"io"
+
+	"github.com/01org/ciao/deploy"
+	"github.com/01org/ciao/payloads"
+	"github.com/01org/ciao/ssntp"
+	"github.com/01org/ciao/ssntp/certs"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+type signingRequest struct {
+	role     ssntp.Role
+	csr      []byte
+	approved bool
+}
+
+var anchorCertPath string
+var caCertPath string
+
+var serverURL string
+var secretToken string
+var autoApprove bool
+var scheduler bool
+var systemPkiLocation string
+var serveRequests bool
+
+// Used when generating the configuration file
+var cephID string
+var httpsCaCertPath string
+var httpsCertPath string
+var identityUser string
+var identityPassword string
+var identityURL string
+var adminSSHKey string
+var adminPassword string
+var computeNet string
+var mgmtNet string
+var serverIP string
+
+// TODO: This may need to be mutex protected. Check if every HTTP request comes in own goroutine.
+var requests map[string]*signingRequest
+
+var ciaoConfigurationDirectory = "/etc/ciao"
+
+func init() {
+	flag.StringVar(&serverURL, "server", ":9000", "Port and optional IP to listen on (host:port)")
+	flag.StringVar(&secretToken, "token", "ciao", "Secret token shared with client")
+	flag.BoolVar(&autoApprove, "auto-approve", true, "Automatically approve requests")
+	flag.BoolVar(&scheduler, "scheduler", true, "Setup scheduler on this machine")
+	flag.StringVar(&systemPkiLocation, "system-pki-location", "/etc/pki/ciao", "Filesystem location for certificates")
+	flag.BoolVar(&serveRequests, "serve-requests", true, "Setup server to listen for join requests")
+	flag.StringVar(&serverIP, "server-ip", "", "IP address that scheduler is reachable on")
+
+	// For configuration file generation
+	flag.StringVar(&cephID, "ceph-id", "ciao", "The ceph id for the storage cluster")
+	flag.StringVar(&httpsCaCertPath, "https-ca-cert", "", "Path to CA certificate for HTTP service")
+	flag.StringVar(&httpsCertPath, "https-cert", "", "Path to certificate for HTTPS service")
+	flag.StringVar(&identityUser, "keystone-user", "", "Username for controller to access keystone (service user)")
+	flag.StringVar(&identityPassword, "keystone-password", "", "Password for controller to access keystone (service user)")
+	flag.StringVar(&identityURL, "keystone-url", "", "URL for keystone server")
+	flag.StringVar(&adminSSHKey, "admin-ssh-key", "", "SSH public key for accessing CNCI")
+	flag.StringVar(&adminPassword, "admin-password", "", "Password for accessing CNCI")
+	flag.StringVar(&computeNet, "compute-net", "", "Network range for compute network")
+	flag.StringVar(&mgmtNet, "mgmt-net", "", "Network range for management network")
+
+	requests = make(map[string]*signingRequest)
+}
+
+func sign(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "PUT" {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	vars := mux.Vars(r)
+	if vars["token"] != secretToken {
```

Comment:
So with this set of tools the token is pre-shared. But the next level of tooling will generate that token so reusing that code will be beneficial.